rust/agents/relayer/src/msg/metadata/base.rs

```diff
@@ -40,39 +49,126 @@ pub struct IsmWithMetadataAndType {
 
 #[async_trait]
 pub trait MetadataBuilder: Send + Sync {
-    #[allow(clippy::async_yields_async)]
     async fn build(&self, ism_address: H256, message: &HyperlaneMessage)
         -> Result<Option<Vec<u8>>>;
 }
 
-#[derive(Clone, new)]
-pub struct BaseMetadataBuilder {
-    destination_chain_setup: ChainConf,
-    origin_prover_sync: Arc<RwLock<MerkleTreeBuilder>>,
-    origin_validator_announce: Arc<dyn ValidatorAnnounce>,
-    allow_local_checkpoint_syncers: bool,
-    metrics: Arc<CoreMetrics>,
-    db: HyperlaneRocksDB,
+/// Allows fetching the default ISM, caching the value for a period of time
+/// to avoid fetching it all the time.
+/// TODO: make this generic
```

Comment:
I wasted a bit of time trying to make this generic where H256 was `T` and there was some update fn in here. Ended up not having much success getting async closures to be able to be passed into here - I'm sure it's possible, @daniel-savu curious if you feel it's worth making this generic atm